google/cloud/internal/big_endian.h

```diff
@@ -15,164 +15,67 @@
-         (((num_value)&0x00ff0000ul) >> 8) | (((num_value)&0x0000ff00ul) << 8) |
-         (((num_value)&0x000000fful) << 24));
-
-    return return_value;
-#endif
-  }
-
-  static inline std::uint64_t byte_swap(std::uint64_t value) {
-#ifdef _MSC_VER
-    return _byteswap_uint64(value);
-#elif defined(__APPLE__)
-    return OSSwapInt64(value);
-#elif defined(__GNUC__) || defined(__clang__)
-    return __builtin_bswap64(value);
-#else
-    std::uint64_t return_value = ((((num_value)&0xff00000000000000ull) >> 56) |
-                                  (((num_value)&0x00ff000000000000ull) >> 40) |
-                                  (((num_value)&0x0000ff0000000000ull) >> 24) |
-                                  (((num_value)&0x000000ff00000000ull) >> 8) |
-                                  (((num_value)&0x00000000ff000000ull) << 8) |
-                                  (((num_value)&0x0000000000ff0000ull) << 24) |
-                                  (((num_value)&0x000000000000ff00ull) << 40) |
-                                  (((num_value)&0x00000000000000ffull) << 56));
-
-    return return_value;
-#endif
+// Encodes an 8-bit, 16-bit, 32-bit, or 64-bit signed or unsigned value as a
+// big-endian sequence of bytes. The returned string has a size matching
+// `sizeof(T)`. Example:
+//
+//   std::string s = EncodeBigEndian(std::int32_t{255});
+//   assert(s == std::string("\0\0\0\xFF", 4));
+//
+template <typename T,
+          typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+inline std::string EncodeBigEndian(T value) {
+  static_assert(std::numeric_limits<unsigned char>::digits == 8,
+                "This code assumes an 8-bit char");
+  using unsigned_type = typename std::make_unsigned<T>::type;
+  unsigned_type const n = value;
+  unsigned_type shift = sizeof(n) * 8;
+  std::string s(sizeof(n), '\0');
+  for (auto& c : s) {
+    shift -= 8;
+    c = (n >> shift) & 0xFF;
   }
+  return s;
+}
 
-  template <typename IntegralType>
-  static inline IntegralType as_big_endian(IntegralType native) {
-    return byte_swap(native);
+// Decodes the given string as a big-endian sequence of bytes representing an
+// integer of the specified type. The allowed types are std::int8_t through
+// std::int64_t, both signed and unsigned. Returns an error status if the given
+// string is the wrong size for the specified type. Example:
+//
+//   std::string s("\0\0\0\xFF", 4);
+//   StatusOr<std::int32_t> decoded = DecodeBigEndian(s);
+//   if (decoded) assert(*decoded == 255);
+//
+template <typename T,
+          typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+inline StatusOr<T> DecodeBigEndian(std::string const& value) {
+  static_assert(std::numeric_limits<unsigned char>::digits == 8,
+                "This code assumes an 8-bit char");
+  if (value.size() != sizeof(T)) {
+    return Status(StatusCode::kInvalidArgument,
+                  "Value must be sizeof(T) bytes long");
```

Comment:
Should we actually include the sizeof(T) value?  By the time the message reaches any eyes, T will be long gone.

Reply:
Done.